node/core/approval-voting/src/criteria.rs

```diff
@@ -483,52 +668,115 @@ pub(crate) fn check_assignment_cert(
 	let public = schnorrkel::PublicKey::from_bytes(validator_public.as_slice())
 		.map_err(|_| InvalidAssignment(Reason::InvalidAssignmentKey))?;
 
-	if claimed_core_index.0 >= config.n_cores {
-		return Err(InvalidAssignment(Reason::CoreIndexOutOfBounds))
+	// Check that we have all backing groups for claimed cores.
+	if claimed_core_indices.count_ones() == 0 ||
+		claimed_core_indices.count_ones() != backing_groups.len()
+	{
+		return Err(InvalidAssignment(Reason::InvalidArguments))
 	}
 
 	// Check that the validator was not part of the backing group
 	// and not already assigned.
-	let is_in_backing =
-		is_in_backing_group(&config.validator_groups, validator_index, backing_group);
+	for (claimed_core, backing_group) in claimed_core_indices.iter_ones().zip(backing_groups.iter())
+	{
+		if claimed_core >= config.n_cores as usize {
+			return Err(InvalidAssignment(Reason::CoreIndexOutOfBounds))
+		}
 
-	if is_in_backing {
-		return Err(InvalidAssignment(Reason::IsInBackingGroup))
+		let is_in_backing =
+			is_in_backing_group(&config.validator_groups, validator_index, *backing_group);
+
+		if is_in_backing {
+			return Err(InvalidAssignment(Reason::IsInBackingGroup))
+		}
 	}
 
-	let vrf_signature = &assignment.vrf;
-	match assignment.kind {
-		AssignmentCertKind::RelayVRFModulo { sample } => {
-			if sample >= config.relay_vrf_modulo_samples {
+	let vrf_output = &assignment.vrf.output;
+	let vrf_proof = &assignment.vrf.proof;
+
+	match &assignment.kind {
+		AssignmentCertKindV2::RelayVRFModuloCompact { core_bitfield } => {
+			// Check that claimed core bitfield match the one from certificate.
+			if &claimed_core_indices != core_bitfield {
+				return Err(InvalidAssignment(Reason::VRFModuloCoreIndexMismatch))
+			}
+
+			let (vrf_in_out, _) = public
+				.vrf_verify_extra(
+					relay_vrf_modulo_transcript_v2(relay_vrf_story),
+					&vrf_output.0,
+					&vrf_proof.0,
+					assigned_cores_transcript(core_bitfield),
+				)
+				.map_err(|_| InvalidAssignment(Reason::VRFModuloOutputMismatch))?;
+
+			let resulting_cores = relay_vrf_modulo_cores(
+				&vrf_in_out,
+				config.relay_vrf_modulo_samples,
+				config.n_cores,
+			);
+
+			// TODO: Enforce that all claimable cores are claimed, or include refused cores.
```

Comment:
Great if possible, but is it?

Reply:
Not really, maybe in the context of https://github.com/paritytech/polkadot/issues/7222 we could slash them for not doing it.

Reply:
I'll remove the TODO anyway.